flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactoryTest.java

```diff
@@ -112,6 +115,35 @@ public void testExceptionsAreNotCached() throws Exception {
         }
     }
 
+    @Test
+    public void testReuseNettyPartitionRequestClient() throws Exception {
+        NettyTestUtil.NettyServerAndClient nettyServerAndClient = createNettyServerAndClient();
+        try {
+            checkReuseNettyPartitionRequestClient(nettyServerAndClient, 1);
+            checkReuseNettyPartitionRequestClient(nettyServerAndClient, 2);
+            checkReuseNettyPartitionRequestClient(nettyServerAndClient, 5);
+            checkReuseNettyPartitionRequestClient(nettyServerAndClient, 10);
+        } finally {
+            nettyServerAndClient.client().shutdown();
+            nettyServerAndClient.server().shutdown();
+        }
+    }
+
+    private void checkReuseNettyPartitionRequestClient(
+            NettyTestUtil.NettyServerAndClient nettyServerAndClient, int maxNumberOfConnections)
+            throws Exception {
+        Set<NettyPartitionRequestClient> set = new HashSet<>();
+
+        final PartitionRequestClientFactory factory =
+                new PartitionRequestClientFactory(
+                        nettyServerAndClient.client(), 0, maxNumberOfConnections);
+        for (int i = 0; i < Math.max(100, maxNumberOfConnections); i++) {
+            final ConnectionID connectionID = nettyServerAndClient.getConnectionID(i);
+            set.add(factory.createPartitionRequestClient(connectionID));
+        }
+        assertEquals(maxNumberOfConnections, set.size());
```

Comment:
Maybe using a random connection index and replace this assertion with ```assertTrue(set.size() <= maxNumberOfConnections);``` which is more like the real case? 
In the current implementation, the selection of connection index value (1 - 100) and the modulus operation (%) guarantee that it can pass, but it seems that this case is not common enough.

Reply:
Hi @wsry , thanks for your review. I have addressed these comments. Please help to review again when you are free. Thanks a lot.

Reply:
@1996fanrui Hi, thanks for your update. Maybe I did not express it clearly which caused misunderstanding. I mean to use random connection index, not random connection index count. Because the real connection index is also randomly generated, you can refer to the IntermediateResult class for more information. In a word, I mean something like this:
```
        final Random random = new Random();
        for (int i = 0; i < Math.max(100, maxNumberOfConnections); i++) {
            final ConnectionID connectionID =
                    nettyServerAndClient.getConnectionID(random.nextInt(Integer.MAX_VALUE));
            set.add(factory.createPartitionRequestClient(connectionID));
        }
        assertTrue(set.size() <= maxNumberOfConnections);
```

Reply:
Got it, thanks for your clarification and reminder. I've got more information from IntermediateResult and resubmit this PR.